sql/src/main/java/io/crate/operation/collect/RowsCollector.java

```diff
@@ -27,48 +27,65 @@
 import io.crate.operation.projectors.RowReceiver;
 
 import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
 
 public class RowsCollector implements CrateCollector {
 
-    private final IterableRowEmitter emitter;
+    private final RowReceiver rowDownstream;
+    private final CompletableFuture<? extends Iterable<?>> future;
+    private final Function<Iterable, Iterable<? extends Row>> transformFunction;
+    private IterableRowEmitter emitter;
 
     public static RowsCollector empty(RowReceiver rowDownstream) {
-        return new RowsCollector(rowDownstream, ImmutableList.<Row>of());
+        return new RowsCollector(
+            rowDownstream,
+            CompletableFuture.completedFuture(ImmutableList.of()),
+            objects -> ImmutableList.of());
     }
 
     public static RowsCollector single(Row row, RowReceiver rowDownstream) {
-        return new RowsCollector(rowDownstream, ImmutableList.of(row));
+        return new RowsCollector(
+            rowDownstream,
+            CompletableFuture.completedFuture(ImmutableList.of(row)),
+            objects -> (Iterable<? extends Row>) objects);
```

Comment:
Maybe rename `objects` to `rows`?
I'm not sure - somehow the term `rows` could refer to both `Iterable<?>` and `Iterable<Row>` which is a bit confusing I guess.

Maybe distinguish via `dataIterable` and `rowIterable` ?